controllers/modelregistry_controller.go

```diff
@@ -0,0 +1,181 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ModelRegistryReconciler holds the model registry controller configuration.
+type ModelRegistryReconciler struct {
+	Client         client.Client
+	Scheme         *runtime.Scheme
+	Log            logr.Logger
+	Period         time.Duration
+	mrISReconciler *reconcilers.ModelRegistryInferenceServiceReconciler
+	mrSEReconciler *reconcilers.ModelRegistryServingEnvironmentReconciler
+}
+
+func NewModelRegistryReconciler(client client.Client, scheme *runtime.Scheme, log logr.Logger, period int) *ModelRegistryReconciler {
+	return &ModelRegistryReconciler{
+		Client:         client,
+		Scheme:         scheme,
+		Log:            log,
+		Period:         time.Duration(period) * time.Second,
+		mrISReconciler: reconcilers.NewModelRegistryInferenceServiceReconciler(client),
+		mrSEReconciler: reconcilers.NewModelRegistryServingEnvironmentReconciler(client),
+	}
+}
+
+func (r *ModelRegistryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Initialize logger format
+	log := r.Log.WithValues("ServingRuntime", req.Name, "namespace", req.Namespace)
+	log.Info("Reconciling ModelRegistry serving for ServingRuntime: " + req.Name)
+
+	// Check if model registry exists in this namespace, lookup the service by label
+
+	opts := []client.ListOption{client.InNamespace(req.Namespace), client.MatchingLabels{
+		"component": "model-registry",
+	}}
+	mrServiceList := &corev1.ServiceList{}
+	err := r.Client.List(ctx, mrServiceList, opts...)
+	if err != nil && apierrs.IsNotFound(err) {
+		// No model registry deployed in the provided namespace, skipping serving reconciliation
+		log.Info("Stop ModelRegistry serving reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if len(mrServiceList.Items) == 0 {
+		log.Info("No Model Registry service found for Namespace: " + req.Namespace)
+		log.Info("Stop ModelRegistry serving reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	// Actually we could iterate over every mrService, as nothing prevents to setup multiple MR in the same namespace
+	if len(mrServiceList.Items) > 1 {
+		log.Error(fmt.Errorf("multiple services with component=model-registry for Namespace %s", req.Namespace), "Stop ModelRegistry serving reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	mrService := mrServiceList.Items[0]
+
+	// Setup model registry service
+
+	// setup grpc connection to ml-metadata
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	var grpcPort *int32
+	for _, port := range mrService.Spec.Ports {
+		if port.Name == "grpc-api" {
+			grpcPort = &port.Port
+			break
+		}
+	}
+
+	if grpcPort == nil {
+		log.Error(fmt.Errorf("cannot find grpc-api port for service %s", mrService.Name), "Stop ModelRegistry serving reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	mlmdAddr := fmt.Sprintf("%s.%s.svc.cluster.local:%d", mrService.Name, req.Namespace, *grpcPort)
```

Comment:
isn't better just to use the service name?

Reply:
Initially I used that but IIUC it does not work cross-namespace, isn't it?
Therefore the only way I found for now is this one but happy to find a better approach :)

Reply:
I see, well, if accessing another namespace is a must, then I think it is the best option as well.
Or if the service is already exposed, that might be a good option as well.

Reply:
The service itself is not exposed, at least for now, and yeah accessing the model registry from different namespaces is a must as we will have one MR for each DS project (if enabled installed obv) at the moment and the `odh-model-controller` should be able to connect to each of them from its own namespace.